editor/src/core/shared/equality-utils.ts

```diff
@@ -71,3 +71,27 @@ export function keepReferenceIfShallowEqual<T>(original: T, maybeNew: T) {
     return maybeNew
   }
 }
+
+function keyEquality<T>(key: keyof T, a: T, b: T): boolean {
+  return a[key] === b[key]
+}
+
+export function keysEquality<T>(keys: ReadonlyArray<keyof T>, a: T, b: T): boolean {
+  return keys.every((key) => keyEquality(key, a, b))
+}
+
+type AtLeastOne<T> = [T, ...T[]]
+
+export const keysEqualityExhaustive =
+  <O extends { [key: string]: any }>() =>
+  <L1 extends AtLeastOne<keyof O>, L2 extends Array<Exclude<keyof O, L1[number]>>>(options: {
+    include: L1
+    exclude: L2 extends any
```

Comment:
What is this field for?

Reply:
its an exhaustive exclude list: the idea is that you must exhaustively enumerate all keys of the tested object: you have to explicitly say if they should be included in the comparison, or ignored. (maybe ignoredKeys would be a better name?)

Reply:
Is this so that if we added another field to the type then it's a compile error until it's added to the include or exclude list?

Reply:
exactly!

Reply:
this is awesome!